### ISSUE TITLE
[question] Package ID unexpected behavior

### ISSUE THREAD

**Issue by Forbinn** [conan-io/conan#8227]
I discovered a weird behaviour in the generation of the package ID for a recipe that requires qt.

Here is the minimal example:

- `conanfile.txt`
```
[requires]
qt/5.15.2@bincrafters/stable
```

Environment:
- docker image `conanio/gcc7`
- `conan profile new --detect default`
- `conan profile update settings.compiler.libcxx=libstdc++11 default`
- `conan remote add bincrafters https://api.bintray.com/conan/bincrafters/public-conan`

Default package ID mode:
- `cat $HOME/.conan/conan.conf | grep default_package_id_mode`
```
default_package_id_mode = semver_direct_mode # environment CONAN_DEFAULT_PACKAGE_ID_MODE
```

If I run `conan info -n id .` I get:
```
conanfile.txt
    ID: a13c4ee5c4526b19b1ae512ddd11c56d93c7e2f2
```

But now If I run `conan info -n id -o qt:with_mysql=False .`, I get:
```
conanfile.txt
    ID: 967981d76f1d4beb7a6e8b10a88a7548c314e7d2
```

Has you can see the package ID of my recipe changes depending on option set on its qt dependencies. The package ID only changes if the option of qt I used actually changes qt recipe own requirements.

Is this behaviour expected ? Because if I look at https://blog.conan.io/2019/09/27/package-id-modes.html, this shouldn't be the case but there is maybe something I missed.

- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @Forbinn 

Yes, the default behavior is taking dependencies options into account. There is a known bug described in https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html, in which these options will affect the consumers package_ids. In general it is kind of expected, the options of the dependencies affect the consumers: if you are building an executable that is statically linking qt, and qt is different depending on whether it implements sql support, then yes, the final executable will be different for different options of ``qt``, won't it?

Please let me know if this clarifies the behavior.

**Comment by Forbinn** [comment]
Hi @memsharded,

Thank you for your reply. Yes you are right some options of an underlying dependency requires a change in the current recipe but this is not always the case, especially with recipe like Qt which contains a lot of options that won't necessarily affect the current one. For example if my library only require the core component of Qt, then adding websockets or sql module will not affect mine.

Also another things that bother me is this:
| Command | Package ID | Qt Package ID |
| - | - | - |
| `conan info -n id .` | `a13c4ee5c4526b19b1ae512ddd11c56d93c7e2f2` | `1b2a7e7a497c985817c8ca6d39aa36755bcebdf8` |
| `conan info -n id -o qt:with_mysql=False .` | `967981d76f1d4beb7a6e8b10a88a7548c314e7d2` | `7e918ffa540dee410697ac7f16ab1a93bf66a055` |
| `conan info -n id -o qt:openssl=False .` | `a13c4ee5c4526b19b1ae512ddd11c56d93c7e2f2` | `96ea24d2872a57be930e97d02e71540d283e422e` |

The Qt package ID is always different which is expected since I'm changing the default value of an option but the package ID of my  recipe sometimes changes and sometimes not which is not what I expected. It should either never change or always change based on what you told me.

Also I thought that the [versioning schema](https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html#versioning-schema) compatibility was here to handle this precisely (but I'm maybe mistaken). If you use the `full_package_mode` then the package ID (and by extension everything that affect that recipe) of the dependency will be taken into account, otherwise it will not.
This is why if I print the `self.info.requires` object in the `package_id` function of my recipe then I only get `qt/5.X.Y`. The package ID of Qt should not be considered.

Can you clarify this?

**Comment by dahakawang** [comment]
Hi @memsharded , G'day!

I've also hit by this issue recently, and first of all I think this "known-bug" exists not only for `semver_direct_mode `, so maybe [the document](https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html#versioning-schema) needs to be more accurate about this.


To me, "*taking dependencies options into account.*" isn't really the biggest problem, the behavioral inconsistency is. Depending on how your recipe's written, Conan sometimes will also overlook dependencies' options while calculating the package_id (I will post example in the following comments)


A fix that makes the behavior consistent could go either ways, the way you mentioned above does make sense to me, but on the other hand we could also argue that a dependency's options is already encoded into its package_id, thus there's no need to look at dependency options while calculating id of the root project. The act of choosing `full_version_mode` vs `full_package_mode` actually reflects the project owner's decision -- on wether he think that a different value to the option of a dependency would result in a rebuild of his project (i.e. package_id change). **So the key is to give user the flexibility** (unless there's already the freedom to tell Conan to ignore/include one dependency option while deciding package_id of the root project?)


So to sum up, "making behavior consistent, and ideally moving away from looking at dependencies' options" is what I hope Conan team could take.

I fully understand the situation as @jgsogo mentioned [in this issue](https://github.com/conan-io/conan/issues/8149#issuecomment-738803363), but I would suggest we at least make the fix as an opt-in (maybe via flags in the recipe) so these who suffers from this known bug and these who rely on it but want to incrementally move away from it could both benefit while the vast majority of existing package remains intact.

What do you guys think?

**Comment by dahakawang** [comment]
bellow follows my example, to demonstrate Conan will & will not have dependency's option participating root project's package id calculation.

I will always have the same root project recipe as below:
```python
from conans import ConanFile
class AppConan(ConanFile):
    name = "my_app"
    version = "1.0"
    requires = "my_lib/1.0"
    def package_id(self):
        self.info.requires['my_lib'].unrelated_mode()
```

If I have below lib project recipe, Conan **WILL** take option with_z into consideration for my_app's package_id:
```python
from conans import ConanFile
class LibConan(ConanFile):
    name = "my_lib"
    version = "1.0"
    options = {"with_z": [True, False]}
    default_options = {"with_z": False}
    def requirements(self):
        if self.options.with_z:
            self.requires("zlib/1.2.11")
```
### result
```bash
conan create conanfile_lib.py -o with_z=False
conan create conanfile_lib.py -o with_z=True
conan info -n id conanfile_app.py -o 'my_lib:with_z=False'   # ==> 0c8b5ebf2790dd989f84360c366965b731a9bfc8
conan info -n id conanfile_app.py -o 'my_lib:with_z=True'    # ==> 2db50bf04a653e81e7735de06a9f6e79c7a712dc
```

However, if I have below lib project recipe, Conan **WILL NOT** take option with_z into consideration for my_app's package_id:
```python
from conans import ConanFile
class LibConan(ConanFile):
    name = "my_lib"
    version = "1.0"
    options = {"with_z": [True, False]}
    default_options = {"with_z": False}
    def requirements(self):
        self.requires("zlib/1.2.11")
```
### result
```bash
conan create conanfile_lib.py -o with_z=False
conan create conanfile_lib.py -o with_z=True
conan info -n id conanfile_app.py -o 'my_lib:with_z=False'   # ==> 2db50bf04a653e81e7735de06a9f6e79c7a712dc
conan info -n id conanfile_app.py -o 'my_lib:with_z=True'    # ==> 2db50bf04a653e81e7735de06a9f6e79c7a712dc
```

**Comment by memsharded** [comment]
``package_id`` computation was completely revamped in Conan 2, closing as outdated. Please open new tickets for any further question or issue, thanks for the feedback!
